### ISSUE TITLE
Extend current plugin model with steps

### ISSUE THREAD

**Issue by ThijsKoot** [schipholgroup/takeoff#65]
### Context
I believe Takeoff would benefit a lot from allowing users to write their own deployment steps. @danielvdende asked me to write an issue to facilitate further discussion on this idea. Plugins would allow for broader use cases not envisioned by Takeoff's original creators without putting pressure on maintainers to add new steps. If community-built steps are mature enough they could be considered for upstream adoption. Similar plugin models already works well for projects like [Apache Airflow](https://airflow.apache.org/docs/stable/plugins.html) and [dbt](https://docs.getdbt.com/docs/contributing/building-a-new-adapter). 

### Proposed solution
The easiest method is allowing users to write steps in the [plugin-folder](https://schipholgroup.github.io/takeoff/takeoff-plugins) by inheriting from `Step`. Plugged steps could be listed in `__init__.py` or `steps.py` in the plugin-folder. The latter option would be consistent with Takeoff's own methodology. The method described here only supports local files and would not promote the use of community-built steps. One possible solution is to allow steps to be installed from PyPI or Github at runtime by specifying a package or repository. 

**Proposed plugin step functionality:**
- Scan plugin folder for steps at runtime
- Steps can extend the base schema
- Plugin steps can be installed from remote sources (PyPI, Github?)
- Plugin steps can access credentials in Key Vault

**Required work**
- Enable use of plugin steps
- Write developer documentation for steps
- Maintain a list of community-provided steps

Let's discuss!

**Comment by danielvdende** [comment]
Thanks for writing this up @ThijsKoot.
I think this is a good idea. I would keep it simple for the first iteration though, and scope it down to only local files (e.g. custom classes inheriting from  `Step` in the plugins folder). I think it would also enhance testability of such steps, by allowing people to use them in their own environment before commiting them to the main project. 
I'm happy to work on this, but also happy to review PRs :slightly_smiling_face:

**Comment by ThijsKoot** [comment]
Agreed on limiting the initial scope, had a few points in my head I wanted to write down. I haven't really looked at how to implement, will do this sometime later this week. I'll see if I can submit an initial PR. Was thinking about getting back to you with a rough idea but that usually translates to PRs in my head so we'll see. 

Sidenote: I already test locally by running the Docker image with the proper environment variables, works really well. Debugging on Gitlab's runners got old real quick.
